File: backend/app/model/schemas/session.py
"""Pydantic models for session API endpoints."""

from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class SessionCreate(BaseModel):
    """Request body for creating a session."""

    title: str = Field("New Conversation", min_length=1, max_length=255)
    edition: Optional[str] = Field(None, description="Default D&D edition (2014/2024)")
    settings: Optional[dict[str, Any]] = Field(None, description="Session-specific settings")


class SessionUpdate(BaseModel):
    """Request body for updating a session."""

    title: Optional[str] = Field(None, min_length=1, max_length=255)
    settings: Optional[dict[str, Any]] = None
    meta_data: Optional[dict[str, Any]] = None


class MessageOut(BaseModel):
    """Message in API responses."""

    id: UUID
    session_id: UUID
    role: str
    content: dict[str, Any]
    created_at: datetime
    sequence_number: int
    provider: Optional[str] = None
    model: Optional[str] = None
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    latency_ms: Optional[int] = None
    stop_reason: Optional[str] = None
    meta_data: dict[str, Any] = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class SessionOut(BaseModel):
    """Session in API responses."""

    id: UUID
    title: str
    created_at: datetime
    updated_at: datetime
    user_id: Optional[str] = None
    settings: dict[str, Any] = Field(default_factory=dict)
    meta_data: dict[str, Any] = Field(default_factory=dict)
    message_count: int = 0

    model_config = {"from_attributes": True}


class SessionDetailOut(SessionOut):
    """Session with messages included."""

    messages: list[MessageOut] = Field(default_factory=list)


class SessionListOut(BaseModel):
    """Paginated session list response."""

    sessions: list[SessionOut]
    total: int
    limit: int
    offset: int

File: frontend/src/types/session.ts
/** Mirrors backend SessionOut / SessionDetailOut / MessageOut schemas. */
import type { ChatToolsMetadata } from "@/types/chat";

export interface MessageMetadata {
	readonly retrieval?: Record<string, unknown> | null;
	readonly timing?: Record<string, unknown>;
	readonly tools?: ChatToolsMetadata;
	readonly [key: string]: unknown;
}

export interface Session {
	readonly id: string;
	readonly title: string;
	readonly created_at: string;
	readonly updated_at: string;
	readonly user_id: string | null;
	readonly settings: Record<string, unknown>;
	readonly meta_data: Record<string, unknown>;
	readonly message_count: number;
}

export interface SessionDetail extends Session {
	readonly messages: Message[];
}

export interface SessionListResponse {
	readonly sessions: Session[];
	readonly total: number;
	readonly limit: number;
	readonly offset: number;
}

export interface SessionCreate {
	readonly title?: string | undefined;
	readonly edition?: string | undefined;
	readonly settings?: Record<string, unknown> | undefined;
}

export interface SessionUpdate {
	readonly title?: string | undefined;
	readonly settings?: Record<string, unknown> | undefined;
	readonly meta_data?: Record<string, unknown> | undefined;
}

export interface Message {
	readonly id: string;
	readonly session_id: string;
	readonly role: "user" | "assistant" | "system";
	readonly content: MessageContent;
	readonly created_at: string;
	readonly sequence_number: number;
	readonly provider: string | null;
	readonly model: string | null;
	readonly prompt_tokens: number;
	readonly completion_tokens: number;
	readonly total_tokens: number;
	readonly latency_ms: number | null;
	readonly stop_reason: string | null;
	readonly meta_data: MessageMetadata;
}

export interface MessageContent {
	readonly text: string;
	readonly sources?: SourceReference[] | undefined;
}

export interface SourceReference {
	readonly file: string;
	readonly content: string;
	readonly score: number;
	readonly line_range?: [number, number] | undefined;
}
